solidity/contracts/BancorNetwork.sol

```diff
@@ -450,6 +405,191 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
+            // otherwise claim the tokens
+            if (isNewerConverter)
+                safeTransferFrom(_sourceToken, msg.sender, firstConverter, _amount);
+            else
+                safeTransferFrom(_sourceToken, msg.sender, this, _amount);
+        }
+    }
+
+    /**
+      * @dev handles the conversion target token if the network still holds it at the end of the conversion
+      * 
+      * @param _data        conversion data, see ConversionStep struct above
+      * @param _amount      conversion return amount, in the target token
+      * @param _beneficiary wallet to receive the conversion result
+    */
+    function handleTargetToken(ConversionStep[] _data, uint256 _amount, address _beneficiary) private {
+        ConversionStep memory stepData = _data[_data.length - 1];
+
+        // network contract doesn't hold the tokens, do nothing
+        if (stepData.beneficiary != address(this))
+            return;
+
+        IERC20Token targetToken = stepData.targetToken;
+
+        // ETH / EtherToken
+        if (etherTokens[targetToken]) {
+            // newer converter should send ETH directly to the beneficiary
+            assert(!stepData.isV27OrHigherConverter);
+
+            // EtherToken converter - withdraw the ETH and transfer to the beneficiary
+            IEtherToken(targetToken).withdrawTo(_beneficiary, _amount);
+        }
+        // other ERC20 token
+        else {
+            safeTransfer(targetToken, _beneficiary, _amount);
+        }
+    }
+
+    /**
+      * @dev creates a memory cache of all conversion steps data to minimize logic and external calls during conversions
+      * 
+      * @param _conversionPath      conversion path, see conversion path format above
+      * @param _minReturn           if the conversion results in an amount smaller than the minimum return - it is cancelled, must be nonzero
+      * @param _beneficiary         wallet to receive the conversion result
+      * @param _affiliateFeeEnabled true if affiliate fee was requested by the sender, false if not
+      * 
+      * @return cached conversion data to be ingested later on by the conversion flow
+    */
+    function createConversionData(IERC20Token[] _conversionPath, uint256 _minReturn, address _beneficiary, bool _affiliateFeeEnabled) private view returns (ConversionStep[]) {
+        ConversionStep[] memory data = new ConversionStep[](_conversionPath.length / 2);
+
+        bool affiliateFeeProcessed = false;
+        address bntToken = addressOf(BNT_TOKEN);
+        // iterate the conversion path and create the conversion data for each step
+        uint256 i;
+        for (i = 0; i < _conversionPath.length - 1; i += 2) {
+            ISmartToken smartToken = ISmartToken(_conversionPath[i + 1]);
+            IBancorConverter converter = IBancorConverter(smartToken.owner());
+            IERC20Token targetToken = _conversionPath[i + 2];
+
+            // check if the affiliate fee should be processed in this step

```

Comment:
I would change this as follows:
1. Outside the loop, declare `uint256 affiliateFeeCount = 0;`
2. Inside the loop, do:
```
if (_affiliateFeeEnabled && targetToken == bntToken)
    affiliateFeeCount += 1;
```
And then:
```
processAffiliateFee: affiliateFeeCount == 1
```
Assuming that the BNT token can occur at most once in any given path, it would yield the same efficiency (gas-cost) with a cleaner code which is easier to understand. In fact, even if the BNT token somehow appears more than once in some input path, it would still yield more or less the same efficiency.

Reply:
nice :)  - done